File: src/lib/toggle-state.ts
import type { Control, ControlState } from '../types';

type ToggleSelectedOptions = boolean | string[];

function getToggleDefaultSelections(control: Control): ToggleSelectedOptions {
  if (Array.isArray(control.initiallySelectedOptions)) {
    return [...control.initiallySelectedOptions];
  }

  if ((control.options?.length ?? 0) > 1) {
    return (control.options ?? []).map((option) => option.id);
  }

  return true;
}

export function createInitialToggleState(control: Control): ControlState {
  if (control.kind !== 'toggle') {
    throw new Error('createInitialToggleState can only be used with toggle controls.');
  }

  const initial = control.initiallySelectedOptions;

  if (Array.isArray(initial)) {
    return {
      selectedOptions: [...initial],
      enabled: false,
      weight: 1,
    };
  }

  if (initial === true) {
    return {
      selectedOptions: getToggleDefaultSelections(control),
      enabled: true,
      weight: 1,
    };
  }

  return {
    selectedOptions: (control.options?.length ?? 0) > 1 ? [] : false,
    enabled: false,
    weight: 1,
  };
}

export function isToggleEnabled(controlState: Pick<ControlState, 'selectedOptions' | 'enabled'>): boolean {
  if (typeof controlState.selectedOptions === 'boolean') {
    return controlState.selectedOptions;
  }

  return controlState.enabled ?? false;
}

export function getToggleSelectionsForNextState(
  control: Control,
  controlState: Pick<ControlState, 'selectedOptions'>,
  enabled: boolean,
): ToggleSelectedOptions {
  if (control.kind !== 'toggle') {
    throw new Error('getToggleSelectionsForNextState can only be used with toggle controls.');
  }

  if (typeof controlState.selectedOptions === 'boolean') {
    return enabled;
  }

  if (!Array.isArray(controlState.selectedOptions)) {
    throw new Error('Toggle controls must use boolean or string[] selectedOptions.');
  }

  if (!enabled) {
    return controlState.selectedOptions;
  }

  // Multi-option toggles keep prior explicit choices; otherwise they fall back to the
  // schema-defined defaults for the first enable action.
  return controlState.selectedOptions.length > 0
    ? controlState.selectedOptions
    : getToggleDefaultSelections(control);
}

File: promptcrafter/toggle_state.py
from promptcrafter.types import Control, ControlState, ManyOf, Selection, Switch


def toggle_holds_an_option_list(control_state: ControlState) -> bool:
    """A toggle whose control offers more than one option narrows to a list;
    one that does not is a plain on/off.

    The two shapes part company three times -- what a flip writes, what
    ``is_toggle_enabled`` reads, and which renderer the kind reaches -- and this
    is the one place that asks which it has.
    """
    return not isinstance(control_state.selected_options, Switch)


def create_initial_toggle_state(control: Control) -> ControlState:
    if control.kind != "toggle":
        raise ValueError("create_initial_toggle_state can only be used with toggle controls.")

    initial = control.initially_selected_options

    if isinstance(initial, list):
        return ControlState(
            selected_options=ManyOf(tuple(initial)),
            enabled=False,
            weight=1,
        )

    if initial is True:
        return ControlState(
            selected_options=_get_toggle_default_selections(control),
            enabled=True,
            weight=1,
        )

    if len(control.options) > 1:
        return ControlState(selected_options=ManyOf(), enabled=False, weight=1)
    return ControlState(selected_options=Switch(False), enabled=False, weight=1)


def is_toggle_enabled(control_state: ControlState) -> bool:
    if not toggle_holds_an_option_list(control_state):
        return control_state.selected_options.is_switched_on()
    return control_state.enabled or False


def get_toggle_selections_for_next_state(
    control: Control,
    control_state: ControlState,
    enabled: bool,
) -> Selection:
    if control.kind != "toggle":
        raise ValueError("get_toggle_selections_for_next_state can only be used with toggle controls.")

    if not toggle_holds_an_option_list(control_state):
        return Switch(enabled)

    if not enabled:
        return control_state.selected_options

    if control_state.selected_options.has_selection():
        return control_state.selected_options
    return _get_toggle_default_selections(control)


def _get_toggle_default_selections(control: Control) -> Selection:
    if isinstance(control.initially_selected_options, list):
        return ManyOf(tuple(control.initially_selected_options))
    if len(control.options) > 1:
        return ManyOf(tuple(opt.id for opt in control.options))
    return Switch(True)
